### ISSUE TITLE
No completions of the functions of SciPy, NumPy, Matplotlib, and PyLab

### ISSUE THREAD

**Issue by changyuheng** [davidhalter/jedi#372]
This issue is very similar to [this](http://stackoverflow.com/questions/14708453/emacs-jedi-does-not-find-numpy-completions).

I'm currently using Vim with YouCompleteMe. I found that Python's code completion went well except of the follow modules' functions: SciPy, NumPy, Matplotlib, and PyLab.

I traced the YouCompleteMe, finding it just pass the code completion job to Jedi, so then gave a quick try:

``` python
import jedi
print(jedi.Script('import numpy; a = numpy.array([0]); a.').completions())
```

Got output: '[]'.

But

``` python
import jedi
print(jedi.Script('import numpy; numpy.').completions())
```

went well.

I also found that IPython completes both 'numpy.' and 'a.' well.

There is a test unit for NumPy in the file 'test/completion/thirdparty/pylab_.py' around the line containing 'na.shape', so I think the root cause may be the wrong configuration on my machine. But I cannot find it out from the documents.

I'm using Vim with +python flag and YouCompleteMe plug-in on Ubuntu 13.10. I also installed SciPy by apt-get:

``` sh
apt-get install python-numpy python-scipy python-matplotlb
```

And then upgrade them by pip (without virtualenv)

``` sh
pip install --upgrdae numpy scipy matplotlib
```

**Comment by Knusper** [comment]
any updates on this?

**Comment by davidhalter** [comment]
No.

**Comment by ronxin** [comment]
@davidhalter Does jedi currently support extracting the return types specified in numpydoc-style docs? @immerrr mentions this at https://github.com/davidhalter/jedi/issues/372#issuecomment-50237535 , and you mentioned that this "is now a part of Jedi", but in https://github.com/davidhalter/jedi/blob/master/jedi/evaluate/docstrings.py I still only see extracting parameters out of numpydocs, and in testing jedi doesn't seem to be tracking the return types (e.g., of matplotlib). Am I missing something?

**Comment by davidhalter** [comment]
@roxin Maybe it's working in the linter branch?

**Comment by ronxin** [comment]
@davidhalter I didn't seem to find it in the linter branch either.

**Comment by Gansgar** [comment]
Why are you just not running a python shell in the background. I know that sounds ambiguous, because there are many cases which could go wrong. But having instead a complete autocompletion is something I would really like. You could also give the option to activate / deactivate it or something like that.

**Comment by bcolsen** [comment]
Pull request #796 can help with this issue.

To recognize ndarray types you need add a couple lines
At line 108 in docstings.py add:

            if p_type == "ndarray":
                p_type = "numpy.ndarray"

So it look like this:

        for p_name, p_type, p_descr in returns:
            if not p_type:
                p_type = p_name
                p_name = ''

            m = re.match('([^,]+(,[^,]+)*?)$', p_type)
            if p_type == "ndarray":
                p_type = "numpy.ndarray"
            if m:
                p_type = m.group(1)
            found.extend(_expand_typestr(p_type))
        return found

I think the problem is ndarray inside numpy is defined in C code.
This doesn't work for numpy.array() or numpy.zeros() but it does for numpy.asarray()

Also plt.Figure() works. but the matplotlib docstrings often don't list returns.  You can add them yourself though.  It's handy for Figure().add_subplot to return Axes for example

**Comment by djangoliv** [comment]
I have a similar problem (but different I think)
```
print(jedi.Script('import numpy; numpy.a').completions())                                                                                                                                                                                
[<Completion: abs>, <Completion: absolute>, <Completion: add>, <Completion: add_newdocs>, <Completion: alen>, <Completion: all>, <Completion: allclose>, <Completion: ALLOW_THREADS>, <Completion: alltrue>, <Completion: allTypes>, <Complet
ion: alterdot>, <Completion: amax>, <Completion: amin>, <Completion: any>, <Completion: arange>, <Completion: arccos>, <Completion: arccosh>, <Completion: arcsin>, <Completion: arcsinh>, <Completion: arctan>, <Completion: arctan2>, <Comp
letion: arctanh>, <Completion: argmax>, <Completion: argmin>, <Completion: argsort>, <Completion: argwhere>, <Completion: around>, <Completion: array>, <Completion: array2string>, <Completion: array_equal>, <Completion: array_equiv>, <Co
mpletion: array_precision>, <Completion: array_repr>, <Completion: array_str>, <Completion: array_type>, <Completion: Arrayterator>, <Completion: as_strided>, <Completion: asanyarray>, <Completion: asarray>, <Completion: asbytes>, <Compl
etion: asbytes_nested>, <Completion: ascontiguousarray>, <Completion: asfarray>, <Completion: asfortranarray>, <Completion: asmatrix>, <Completion: asscalar>, <Completion: asstr>, <Completion: atleast_1d>, <Completion: atleast_2d>, <Comp
letion: atleast_3d>, <Completion: AxisConcatenator>]
```
doesn't contains average method.

Regards

**Comment by bcolsen** [comment]
I cannot reproduce this on Anaconda 4.2
```
print(jedi.Script('import`` numpy; numpy.a').completions())
[<Completion: abs>, <Completion: absolute>, <Completion: absolute_import>,
<Completion: add>, <Completion: add_docstring>, <Completion: add_newdoc>, 
<Completion: add_newdoc_ufunc>, <Completion: add_newdocs>, <Completion: alen>, 
<Completion: all>, <Completion: allclose>, <Completion: ALLOW_THREADS>, 
<Completion: alltrue>, <Completion: allTypes>, <Completion: alterdot>, 
<Completion: amax>, <Completion: amin>, <Completion: angle>, <Completion: any>, 
<Completion: append>, <Completion: apply_along_axis>, <Completion: apply_over_axes>, 
<Completion: arange>, <Completion: arccos>, <Completion: arccosh>, <Completion: arcsin>,
 <Completion: arcsinh>, <Completion: arctan>, <Completion: arctan2>, <Completion: arctanh>, 
<Completion: argmax>, <Completion: argmin>, <Completion: argpartition>, 
<Completion: argsort>, <Completion: argwhere>, <Completion: around>, <Completion: array>, 
<Completion: array2string>, <Completion: array_equal>, <Completion: array_equiv>, 
<Completion: array_precision>, <Completion: array_repr>, <Completion: array_split>, 
<Completion: array_str>, <Completion: array_type>, <Completion: Arrayterator>, 
<Completion: as_strided>, <Completion: asanyarray>, <Completion: asarray>, 
<Completion: asarray_chkfinite>, <Completion: asbytes>, <Completion: asbytes_nested>, 
<Completion: ascontiguousarray>, <Completion: asfarray>, <Completion: asfortranarray>, 
<Completion: asmatrix>, <Completion: asscalar>, <Completion: asstr>, 
<Completion: atleast_1d>, <Completion: atleast_2d>, <Completion: atleast_3d>, 
<Completion: average>, <Completion: AxisConcatenator>]
```
Versions
python 3.5
```
>>> import numpy
>>> numpy.__version__
'1.11.3'
>>> import jedi
>>> jedi.__version__
'0.9.0'
```

**Comment by oracleyue** [comment]
@bcolsen Yes! It seems that the bug has been fixed in newest `jedi`. I can properly auto-complete `numpy.linalg.d` etc. Seems no problem now.

**Comment by bcolsen** [comment]
@oracleyue this bug is about this code not completing:
```
import jedi
print(jedi.Script('import numpy; a = numpy.array([0]); a.').completions())
```
This still doesn't work. Howerver, this is fixed by applying PR: #796 and PR: #818 

If anyone is interested this issue is patched for Spider-IDE 3.1

![spyder](https://cloud.githubusercontent.com/assets/10513354/22381914/be7a11e0-e480-11e6-8cc9-8f59c1f2ef82.jpg)

**Comment by oracleyue** [comment]
Regarding the bug proposed here: what `numpy` originally defined and strongly recommended is `ndarray`; `numpy.array` is something really strange. 
If you use `a = numpy.ndarry([0,1]); a.`,  it works well. For `array`, no idea, but IPython works well. 
<img width="345" alt="screen shot 2017-01-27 at 19 47 55" src="https://cloud.githubusercontent.com/assets/5046605/22383320/8e2d8d84-e4c9-11e6-9e01-84baba3a7751.png">

Just brainstorm: is it possible to ask IPython developers to get the idea on how they do completions for `numpy,scipy` etc.?

---------
*you might ignore the following; I just suddenly want to mention it. I thought it has been fixed.*

I found one more bug (It's really due to the strange organization of `numpy` `scipy` and their "inheritance" relations, I guess.)
```
import jedi
print(jedi.Script('import numpy; import scipy; import scipy.linalg; scipy.linalg.i').completions())
```
the output is empty. 
However, it should give some completions, e.g. as in ipython:
```
import numpy
import scipy
import scipy.linalg
```
when you use `scipy.linalg.i`, you will get 3 candidates.

But, if you use `import scipy.linalg as linalg; linalg.i`, `jedi` can give correct completion. I put this bug before. Now I have kind of give up; no hope unless diving into the organizations of `numpy, scipy`. I just changed my way to import modules.

FYI,
IPython 3.1.0
Python 2.7.11
Scipy 0.18.0
Numpy 1.12.0

**Comment by bcolsen** [comment]
Those are some good workarounds, but you would have to wrap all array creation routines in np.ndarray to get completions like `np.ndarray(np.zeroes(2,3))`
```
>>> print(jedi.Script('import numpy; a = numpy.ndarray(numpy.zeros(10)); a.a').completions())
[<Completion: all>, <Completion: any>, <Completion: argmax>, <Completion: argmin>, <Completion: argpartition>, <Completion: argsort>, <Completion: astype>]
>>> print(jedi.Script('import numpy; a = numpy.zeros(10); a.a').completions())[]
```
but this is fixed by applying PR: #796 and PR: #818

IPython has it easy...You are constantly executing your code. 
From davidhalter earlier in this thread 
>It's very easy for IPython to understand it, because you can just do a dir() on any object in the REPL. It's way harder to do that without execution of code (which has already been done by the user at that point in IPython).
>Even Python's readline.parse_and_bind("tab: complete") understands all of these. However, Jedi is of a different nature. Jedi does static analysis. That's way harder to get right. We need both. (And as a side note: Jedi's REPL completion also understands numpy and so forth. But in that case, Jedi's also using the Python objects that are available.)

**Comment by davidhalter** [comment]
AFAIK the numpydoc fixes have been applied. You need to install numpydoc to actually have completions. Just sayin'. :)

If you still have an issues (I'm sure there are with this numpy beast), please open a new issue, this one is old and a lot of things have been fixed/changed.
